### PR TITLE
[operator] Improve `Gardenlet` watch handling

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area ops-productivity
/kind bug

**What this PR does / why we need it**:
This PR improves the Gardenlet controller registration in `gardener-operator`. The access token (secret `gardener` or `gardener-internal`) is regularly renewed because it's only valid for `12h`.

When a token expired, that was initially used to start the controller, the watch interrupted with the following error:

```
Failed to watch *v1alpha1.Gardenlet: the server has asked for the client to provide credentials
```

Only a restart of `gardener-operator` helped to recover from this situation.

**Special notes for your reviewer**:
/cc @rfranzke thanks for the inspiration 🙂 

**Release note**:

```bugfix operator
An issue was fixed in `gardener-operator` that led to an inactive Gardenlet controller after a certain period. Thus, the operator needed a restart to react on Gardenlet resources.
```

### CODE REVIEW

File: pkg/operator/controller/controllerregistrar/reconciler.go
Comment (LucaBernstein) [inline pkg/operator/controller/controllerregistrar/reconciler.go]:
```suggestion
// RequeueAfter is the time the request is requeued for in case a controller is not added yet. Exposed for testing.
```

File: docs/concepts/operator.md
Comment (LucaBernstein) [inline docs/concepts/operator.md]: I currently find it hard to understand what is meant by this sentence (managed by gardenlet or managing gardenlet?)
Maybe rephrasing would help...

File: pkg/operator/controller/add.go
Comment (LucaBernstein) [inline pkg/operator/controller/add.go]:
```suggestion
					log.Info("Service DNS name lookup of virtual-garden-kube-apiserver successful, using internal kubeconfig", "error", err)
```

**Review by LucaBernstein — COMMENTED** [review]
lgtm, just some minor wording suggestions...

File: docs/concepts/operator.md
Comment (rfranzke) [inline docs/concepts/operator.md]:
```suggestion
This reconciler performs actions related to the garden access secret (`gardener` or `gardener-internal`) for the virtual garden cluster.
```

File: docs/concepts/operator.md
Comment (rfranzke) [inline docs/concepts/operator.md]:
```suggestion
It extracts the included Kubeconfig, and prepares a dedicated REST config, where the inline bearer token is replaced by a bearer token file with containing said token.
```

File: docs/concepts/operator.md
Comment (rfranzke) [inline docs/concepts/operator.md]:
```suggestion
Afterwards, this REST config is used to add the [`Gardenlet` controller](#gardenlet-controller). 
```

File: docs/concepts/operator.md
Comment (rfranzke) [inline docs/concepts/operator.md]:
```suggestion
> If the garden cluster is a seed cluster at the same time, `gardenlet` will skip running the `NetworkPolicy` and `VPA EvictionRequirements` controllers to avoid interferences.
```

File: pkg/operator/controller/garden/access/add.go
Comment (rfranzke) [inline pkg/operator/controller/garden/access/add.go]:
```suggestion
const ControllerName = "virtual-cluster-access"
```

to make it more clear?

File: pkg/operator/controller/garden/access/add.go
Comment (rfranzke) [inline pkg/operator/controller/garden/access/add.go]: Why OS? Can't we work with the in-memory file system? This way, we don't need to persist actual files anywhere?

File: pkg/operator/controller/garden/access/add.go
Comment (rfranzke) [inline pkg/operator/controller/garden/access/add.go]: Nit: This fits well into a single line, no?

File: pkg/operator/controller/garden/access/add.go
Comment (rfranzke) [inline pkg/operator/controller/garden/access/add.go]: Nit: Typically, we put such predicates into a single function that can be unit-tested.

File: pkg/operator/controller/garden/access/reconciler.go
Comment (rfranzke) [inline pkg/operator/controller/garden/access/reconciler.go]: Similar comments as in the docs

File: pkg/operator/controller/garden/access/reconciler.go
Comment (rfranzke) [inline pkg/operator/controller/garden/access/reconciler.go]:
```suggestion
	kubeConfigRaw, ok := secret.Data[kubernetesclient.KubeConfig]
```

Nit/often called like this

File: pkg/operator/controller/garden/access/add.go
Comment (rfranzke) [inline pkg/operator/controller/garden/access/add.go]: Wouldn't it make more sense to check here if the `token` key is populated in the `data`? You check this anyways in the reconciler, but I guess it's more appropriate here in the predicate already.

File: pkg/operator/controller/garden/access/reconciler.go
Comment (rfranzke) [inline pkg/operator/controller/garden/access/reconciler.go]:
```suggestion
	Config  *config.OperatorConfiguration
	FS            afero.Fs

	tokenFilePath string
	gardenletControllerAdded bool
```

File: pkg/utils/gardener/gardener_status_test.go
Comment (rfranzke) [inline pkg/utils/gardener/gardener_status_test.go:1]: Rename file to `garden_status_test.go`

File: test/integration/operator/garden/access/access_suite_test.go
Comment (rfranzke) [inline test/integration/operator/garden/access/access_suite_test.go]: MemFs?

File: test/integration/operator/garden/access/access_test.go
Comment (rfranzke) [inline test/integration/operator/garden/access/access_test.go]: Nit: block not needed

File: pkg/operator/controller/controllerregistrar/reconciler.go
Comment (rfranzke) [inline pkg/operator/controller/controllerregistrar/reconciler.go]: It's exposed for testing, but I don't see a test?

File: pkg/operator/controller/controllerregistrar/reconciler.go
Comment (rfranzke) [inline pkg/operator/controller/controllerregistrar/reconciler.go:29]: Wouldn't it be cleaner if this returned `(reconcile.Result, error)` like the `Reconcile` itself? Then we also don't need a global `RequeueAfter` constant.

File: pkg/operator/controller/add.go
Comment (rfranzke) [inline pkg/operator/controller/add.go]: I don't think we need this complexity anymore, do we? Since a longer time, we don't support controllers like gardenlet, GOP, etc., to run outside of the cluster. Wouldn't it be enough to just wait for the `Garden` reconciliation?

If you still want to keep it, then make it at least re-useable instead of duplicating the garden clientmap code.

File: pkg/operator/controller/garden/access/reconciler.go
Comment (rfranzke) [inline pkg/operator/controller/garden/access/reconciler.go]: I would not have expected this to happen in this controller. While I understand that this (probably motivated by pragmatic reasons) choice, wouldn't it be cleaner if this was done by the controller registrar as well? I have the following in mind:

- This controller writes the token to a file in the in-memory file system. Not to a temporary file, but to a file with a well-known/constant path (known in this code).
- The controller registrar gets a new entry which checks if this file exists, and if yes, creates the REST config and the `virtualCluster` object which gets added to the manager. 
- Another new entry in the controller registrar checks if the `virtualCluster` object was initialized, and then adds the controllers which need it (atm, only the `Gardenlet` controller).
- We could even augment the `AddToManagerFunc` signature to not only pass the `garden` object, but also the `virtualCluster`.

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/operator/controller/garden/access/add.go
Comment (timuthy) [reply pkg/operator/controller/garden/access/add.go]: Unfortunately, no. `client-go` again reads the file with the `os` package and the in-memory FS is not available there.

File: pkg/operator/controller/garden/access/add.go
Comment (timuthy) [reply pkg/operator/controller/garden/access/add.go]: Hm, is it really? I used this check here because it's very cheap compared to extracting and unmarshalling the kubeconfig. 

I even think it's fine to ignore this in the predicate (hence, remove it) and only do it in the reconciler, where we need the token anyway. It's only relevant at the very beginning of the Garden creation.

File: pkg/operator/controller/garden/access/reconciler.go
Comment (timuthy) [reply pkg/operator/controller/garden/access/reconciler.go]: `gardenletControllerAdded` is exposed for the integration test where we actually skip adding the Gardenlet controller.

File: pkg/operator/controller/garden/access/add.go
Comment (rfranzke) [reply pkg/operator/controller/garden/access/add.go]: :( Fair enough. Then we have not much choice.

File: pkg/operator/controller/garden/access/add.go
Comment (rfranzke) [reply pkg/operator/controller/garden/access/add.go]: I missed that this access secret does not only contain a token, but a full kubeconfig, and the token is populated directly into it. In this light, of course, you're right - let's keep it like this for pragmatic reasons.

File: pkg/operator/controller/garden/access/reconciler.go
Comment (rfranzke) [reply pkg/operator/controller/garden/access/reconciler.go]: See comments below which might make this field obsolete anyhow.

**Comment by timuthy** [comment]
Thanks for your review @rfranzke. I will address your comments and the points from our discussion next week.

File: pkg/operator/controller/controllerregistrar/reconciler.go
Comment (timuthy) [reply pkg/operator/controller/controllerregistrar/reconciler.go:29]: It didn't feel like a natural choice for me to return `(reconcile.Result, error)` when not being directly in a `Reconciler`.
I was more looking for a way to gracefully re-enqueue in case a `AddToManagerFunc` needs to wait longer (similar to or inspired by `retry.Until`), with a common `RequeueAfter`.

File: pkg/operator/controller/add.go
Comment (timuthy) [reply pkg/operator/controller/add.go]: As discussed, I extracted it to a separate function.

File: pkg/operator/controller/garden/access/reconciler.go
Comment (timuthy) [reply pkg/operator/controller/garden/access/reconciler.go]: Please see a139fb1fad309adebf57b4f6e72bed1c0759fffe. I skipped the second step and let the access controller also create the virtual cluster object. Hope that's fine.

**Comment by oliver-goetz** [comment]
Nice improvement 🚀 
/lgtm

File: pkg/operator/controller/garden/access/add.go
Comment (rfranzke) [inline pkg/operator/controller/garden/access/add.go]: If `r.Manager` is set, then `r.Client` should probably be set to `r.Manager.GetClient()`.

File: pkg/operator/controller/garden/access/reconciler.go
Comment (rfranzke) [inline pkg/operator/controller/garden/access/reconciler.go]: Only pass the `VirtualClientConnection` instead of the full config?

File: pkg/operator/controller/garden/access/reconciler.go
Comment (rfranzke) [inline pkg/operator/controller/garden/access/reconciler.go]: I guess we discussed this some weeks back, but I forgot again. Can you please explain (and augment the comment) why we need this (and cannot just use the `MemOs` from `afero` for testing)?

Edit: Found https://github.com/gardener/gardener/pull/10663#discussion_r1804611575. While I understand that we have to use the real FS due to client-go in the production case, we could at least use the in-memory FS for the tests, or? Then we could remove this `CreateTemporaryFile` and just pass a `MemOs` into the unit tests? Maybe I'm still missing something.

File: pkg/operator/controller/garden/access/reconciler.go
Comment (rfranzke) [inline pkg/operator/controller/garden/access/reconciler.go]: I really don't want to sabotage you, but I still find it a but unintuitive that this controller creates the `cluster.Cluster` object. I see and agree that it is simple and straight-forward, but as reader of the code, I wouldn't expect it here.
What I have in mind is that we decide on a temporary file path before instantiating the `garden-access` controller, and pass this path to it. This way, another controller could just wait for the token to be written to the path, and then create and add the `cluster.Cluster`.

Do you agree that this makes sense conceptually or do you see any difficulties/downsides with it? I leave it up to you if you want to follow-up with it. We could also consider it for a subsequent improvement/PR after this one got merged.

File: pkg/operator/controller/add.go
Comment (rfranzke) [inline pkg/operator/controller/add.go]:
```suggestion
					log.Info("Garden is being reconciled - adding the access reconciler will be tried again")
```

File: pkg/operator/controller/add.go
Comment (rfranzke) [inline pkg/operator/controller/add.go]: Can you adapt this error message? It's not only about the NP registrar anymore

**Comment by rfranzke** [comment]
@timuthy I've set this PR to "draft".

During your absence, I tried out the approach I proposed in https://github.com/gardener/gardener/pull/10663#discussion_r1854192257. I've pushed a few new commits to your branch/this PR (prefixed with `[rf]`), but the most relevant one is https://github.com/gardener/gardener/pull/10663/commits/a4d41b7722b388565b5f17fc9c58d20b21458396. Please take a look.

Naturally, it is slightly more code, but I like the separation. As part of the exercise, I also improved the directory structure in the `controller` package, and the logging/handling in the `controllerregistrar`. The unit/integration tests still have to be adapted/added.

I would like to get your opinion on this. We can also have call if you prefer :)

File: pkg/operator/controller/garden/access/reconciler.go
Comment (timuthy) [reply pkg/operator/controller/garden/access/reconciler.go]: As discussed offline, we will follow the suggested enhancement 🙂

File: pkg/operator/controller/garden/access/reconciler.go
Comment (timuthy) [reply pkg/operator/controller/garden/access/reconciler.go]: I adjusted the test to use `MemOs`.  `CreateTemporaryFile` is still in place to influence the file path which is otherwise unknown for temporary files - see 2dc1b177808da7b029d247322f29071087fc92c5.

**Comment by timuthy** [comment]
@rfranzke the PR is ready again. PTAL.

**Review by rfranzke — COMMENTED** [review]
Awesome @timuthy, thanks for picking up my proposal and completing this :)
Let's finally go!

/lgtm
/approve

**Pull request merged**
